File: 525.cpp
#include <unordered_map>
#include <algorithm>
using namespace std;

class Problem525 {
public:
    int findMaxLength(vector<int>& nums) {
        unordered_map<int, int> map = {{0, -1}};
        int len = 0, sum = 0;
        
        for (int i = 0; i < nums.size(); i++){
            (nums[i] == 0) ? sum-- : sum++;
            (map.count(sum)) ? len = max(len, i-map[sum]) : map[sum] = i;
        }
        
        return len;
    }
};

File: 525.py
class Solution(object):
    def findMaxLength(self, nums):
        found = dict()
        found[0] = -1
        s = maxLen = 0
        for i, num in enumerate(nums):
            s = s + 1 if num else s - 1
            if s in found: maxLen = max(maxLen, i - found[s])
            else: found[s] = i
        return maxLen